pkg/agent/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -442,6 +515,10 @@ loop:
 		klog.Errorf("Failed to handle init events: %v", err)
 		return
 	}
+	if !w.restartSynced {
+		// Notify restartFullSyncGroup that all events before bookmark is handled
+		w.restartFullSyncGroup.Done()
```

Comment:
This code block is always executed after the bookmark events are all received after the watcher is restarted (L495-L517). The difference is that we only want to notify the restartFullSyncGroup if it's an agent restart case; if agent has been running and the watchers are re-connected, we do not notify the restartFullSyncGroup. Does that make sense?

Reply:
This is a valid point, thanks for the catch. I think this can be fixed by setting restartSynced to true for all watchers as soon as the three restartFullSyncGroup are all notified in the Run() function for controller. Code updated.

Reply:
This is a valid point, thanks for the catch. I think this can be fixed by setting restartSynced to true for all watchers as soon as the three restartFullSyncGroup are all notified in the Run() function for controller. Code updated.

Reply:
I think there is no harm in calling `Done` multiple times even with a single `Add(1)` call. So maybe given that we could just remove the `restartSynced` flag and call `w.restartFullSyncGroup.Done` unconditionally, as it would make the code simpler.  I think that the race condition that @wenyingd mentioned is still possible in theory (watcher calls `Done()`, there is a disconnection right after that, the watcher executes this if statement after reconnection just as the `Run` function sets the flag to true).

Reply:
waitGroup does not support this: `panic: sync: negative WaitGroup counter`. For the race condition, yes theoretically the scenario *could* happen, but I don't think the watcher can go though reconnecting -> handling the init events quicker than the three Wait() calls.... or are you talking about the case where one watcher called Done() and reconnects, the next time another watcher called Done() and reconnects, and eventually the three waitGroups are all unblocked while the watchers are still re-connecting?

Reply:
Sorry for the WaitGroup, I didn't check carefully enough. I know this is probably unlikely but I'd rather not be in a situation where it could happen at all, as it is easy enough to fix it now, but would be super hard to troubleshoot in the future if it happened.

Looking at the code more, can't you just have this (instead of setting the flags in `Run`)?
```
	if !w.restartSynced {
                 w.restartSynced = true
		// Notify restartFullSyncGroup that all events before bookmark is handled
		w.restartFullSyncGroup.Done()
	}
```
Since there is only one goroutine per watcher, it seems that it would avoid the race condition and work as you intend it to work. Am I missing something?

Reply:
Thanks for the suggestion. Refactored it this way.

Reply:
The alternative would be to use a channel? But I don't see it reduces the complexity here? i.e. a separate channel needs to be populated to each watcher, and each watcher notify the channel when full sync completes

Reply:
I think Quan is suggesting that you use a single `WaitGroup` instead of 1 per watcher. Instead of calling `Add(1)` on each group, you will call `Add(3)` on this one group.

Reply:
@antoninbas This was my original implementation, but Quan brought up a case where `a watcher is restarted multiple times while the others never connect to API successfully`, so in that case the appliedToGroup watcher would have called Done() three times and unblock the waitGroup while the other watchers are not connected

Reply:
I thought you would still have the boolean flag per watcher to prevent each one from calling `Done` multiple times?

Reply:
Right makes sense...... Updated this to one waitGroup